pkg/reader/ccip.go

```diff
@@ -477,124 +492,145 @@ func (r *ccipChainReader) GasPrices(ctx context.Context, chains []cciptypes.Chai
+	resp := map[string]map[cciptypes.ChainSelector][]byte{
+		consts.ContractNameOnRamp: onramps,
+		consts.ContractNameNonceManager: {
+			destChain: staticConfig.NonceManager,
+		},
 	}
+	return resp, nil
+}
 
-	if staticConfig.ChainSelector != r.destChain {
-		return fmt.Errorf("invalid configuration detected, somehow reading from non-dest offramp")
+// bindReaderContract is a generic helper for binding contracts to readers, the addresses input is the same object
+// returned by DiscoverContracts.
+//
+// No error is returned if contractName is not found in the contracts. This allows calling the function before all
+// contracts are discovered.
+//
+//nolint:unused // it will be used soon.
+func (r *ccipChainReader) bindReaderContract(
+	ctx context.Context,
+	chainSel cciptypes.ChainSelector,
+	contractName string,
+	address []byte,
+) error {
+	if err := r.validateReaderExistence(chainSel); err != nil {
+		return fmt.Errorf("validate reader existence: %w", err)
 	}
 
-	// Bind the nonceManager contract address to the reader.
+	encAddress := typeconv.AddressBytesToString(address, uint64(chainSel))
+
+	// Bind the contract address to the reader.
 	// If the same address exists -> no-op
 	// If the address is changed -> updates the address, overwrites the existing one
-	// If the contract not binded -> binds to the new address
-	if err := destReader.Bind(ctx, []types.BoundContract{
+	// If the contract not bound -> binds to the new address
+	if err := r.contractReaders[chainSel].Bind(ctx, []types.BoundContract{
 		{
-			Address: typeconv.AddressBytesToString(staticConfig.NonceManager, uint64(r.destChain)),
-			Name:    consts.ContractNameNonceManager,
+			Address: encAddress,
+			Name:    contractName,
 		},
 	}); err != nil {
-		return fmt.Errorf("bind nonce manager: %w", err)
+		return fmt.Errorf("unable to bind %s for chain %d: %w", contractName, chainSel, err)
 	}
 
 	return nil
 }
 
-func (r *ccipChainReader) Sync(ctx context.Context) (bool, error) {
-	// Note: offramp must be bound first, otherwise onramp bindings
-	// will fail.
-	if err := r.bindOfframp(ctx); err != nil {
-		return false, err
+// newSync goes through the input contracts and binds them to the contract reader.
+//
+//nolint:unused // it will be used soon.
+func (r *ccipChainReader) newSync(ctx context.Context, contracts ContractAddresses) error {
+	var errs []error
+	for contractName, chainSelToAddress := range contracts {
+		for chainSel, address := range chainSelToAddress {
+			// try to bind
+			err := r.bindReaderContract(ctx, chainSel, contractName, address)
+			if err != nil {
+				if errors.Is(err, ErrContractReaderNotFound) {
+					// don't support this chain
+					continue
+				}
+				// some other error, gather
+				// TODO: maybe return early?
+				errs = append(errs, err)
+			}
+			// error is nil, nothing to do
+		}
 	}
+	return errors.Join(errs...)
 
-	if err := r.bindOnramps(ctx); err != nil {
-		return false, err
-	}
+	// OffRamp
```

Comment:
Commented out code still needed?

Reply:
I'll revisit this in a future PR when enabling these functions in the integration test.